### DISCUSSION TITLE
Handwriting - drawing - stylus support --> Does it exist? Would it be developable?

### DISCUSSION THREAD

**Discussion by andrebaerlocher** [TiddlyWiki/TiddlyWiki5#7481] [Ideas]
Dear all, hoping this is the right place to ask:
Is there a plugin that allows (functional) handwriting or drawing with a stylus? To my knowledge, there is a plugin for drawing with limited functionality.

If not, is it possible to develop? And if yes, is this the place to hire someone?

**Comment by pmario** [comment]
There is a very big difference between "handwriting" and "drawing" in my personal understanding. .. So I think it's important to define the terms to make it possible to "calculate" the effort needed to implement something. 

**Drawing**

1. Create pixel images like PNG, JPG or GIF 
2. Create vector images like SVGs

**Writing**

3. Using a "drawing" canvas to collect text input that can be transformed into plain text immediately or later

------

Add 1) Pixel Images

TW allows us to create a very basic drawing canvas that creates PNGs as the following screen shot shows. 

BUT There is 1 important disadvantage: **Embedded images can become huge, in terms of HTML file size.** 

The image below has about 27kByte

![grafik](https://github.com/Jermolene/TiddlyWiki5/assets/374655/9bf2b193-bd8e-45b6-80fb-dd161bca21e7)

The following image has 54kByte. ... It does not seem much, but there is almost no content in the image. 
It could be described with text like so. 

>The image contains 3 "sketchy" houses with 2 scribbled lines. One below the houses one on the right side.
The houses have 3 colours and 2 different sizes. The left small house is pink. The middle bigger house is back and the right house is plum colour. The right scribbled line is dark-green and the line below is plum

This text is 1 kb, which is the smallest file size my file system can handle. 

![grafik](https://github.com/Jermolene/TiddlyWiki5/assets/374655/8d8de28b-fd95-488a-9ae0-11a88298e442)

-----

Add 2) SVG images.

TW uses SVG images quite a lot, but we use external tools like eg: InkScape to create and edit them. Implementing SVG drawing tools into TW imo is not worth it. There are a lot of OpenSource tools which do a much better job and they are ready to be used. 

----------

Add 3) Using a drawing canvas to "collect" text messages. IMO TW is _not_ the right tool for the job. .. We could implement a convert "handwritten text" to "text" engine, but it would be a huge effort and the quality would be poor. 

It's much easier to do it with my phone. As you can see, I did write 2 lines of German text on a sheet of paper, using extra "ugly" handwriting.
It was detected perfectly. The only "problem" was the second "Test" where it detected "Terl" ... But to be honest .. It was right. 

So could we implement something like this into TW. .. No
Can we use it ... yes. Copy the text to the clipboard and paste it into a tiddler. 

**Conclusion**

I think **Using the right tool for the job** is the way to go. .. 

![grafik](https://github.com/Jermolene/TiddlyWiki5/assets/374655/d360aa4f-4f3a-4633-bb69-5e85141eee98)

For much more text I would switch to the "Live Transcribe" tool, which was one of the main reasons, why I bought my current phone.

**Reply by Jermolene** [reply]
@pmario I think that this answer is problematic. The original request is reasonable, but your response questions the validity of the request.

**Reply by pmario** [reply]
>...The original request is reasonable, but your response questions the validity of the request.

I did not want to question the validity of the request. 

In 1) and 2) I think, I did show what can be achieved with the existing possibilities. If that's enough it is OK and it is there. 

With 3) I wanted to show what I, as a user, expect from a function that allows me to take hand written notes on a graphical canvas. 
In a second step, I do want to convert it to plain text. 

I wanted to point out, that implementing this functionality in a similar quality as it exist on modern devices is hard to achieve in a self contained single page app as TW without "blowing the file size through the roof".

I think, at the moment "handwritten text" to "plain text" conversion in TW, without 3rd party online based tools is hard to achieve.

**Reply by andrebaerlocher** [reply]
I think pmario asked a good question: I did not mean handwriting recognition, I meant basically to draw on a canvas. So much is true.

I'd like to have an alternative to OneNote where I could write and draw on a canvas, basically. I have since found slate-write, which would allow to easily write something without clicking edit, waiting for it to load, then saving it.

So the remaining question would be whether it would be possible to draw on a tiddler that is being edited and worked on with slate-write?

**Reply by anicolao** [reply]
@andrebaerlocher can you link the specific `slate-write` thing you mean? When I tried to search for it I found a markdown editor that didn't seem to have what you want.

If I understand what you want, you'd like to be able to draw notes or write notes with a stylus in a tiddler instead of typing the content. I am very interested in these drawing type cases (I own a ReMarkable and use Good Notes on an iPad all the time) but it's a big job to do it well. I might be interested in implementing something minimal that could suit your needs if I can understand them properly.

**Comment by Jermolene** [comment]
> Dear all, hoping this is the right place to ask: Is there a plugin that allows (functional) handwriting or drawing with a stylus? To my knowledge, there is a plugin for drawing with limited functionality.

Hi @andrebaerlocher TiddlyWiki 5 includes a basic sketching facility as @pmario notes. It works well on machines equipped with a stylus, and handwriting works too.

> If not, is it possible to develop? And if yes, is this the place to hire someone?

What functionality are you looking for over and above the existing facility?

**Reply by andrebaerlocher** [reply]
Good question, my initial request was a bit wonky, at best:
I'd like to have an alternative to OneNote where I could write and draw on a canvas, basically. I have since found slate-write, which would allow to easily write something without clicking edit, waiting for it to load, then saving it.

So the remaining question would be whether it would be possible to draw on a tiddler that is being edited and worked on with slate-write?

**Reply by linonetwo** [reply]
[Slate-write](https://tw-cpl.netlify.app/#linonetwo%2Fslate-write) is a WYSIWYG editor, and [tiddlywiki-whiteboard](https://tw-cpl.netlify.app/#linonetwo%2Ftw-whiteboard) is the drawing plugin, maybe you misremembered them?

**Reply by linonetwo** [reply]
Loading widgets like whiteboard is WIP in slate-write, I don't have much time for it because I'm busy writing AI related things in TidGi, and writing game in my startup.
